### PR TITLE
ALPN Tokens and Alt-Svc

### PR DESCRIPTION
**Pull request by MikeBishop**
For #87 -- adds ALPN token guidelines and IANA registration.  Also defines Alt-Svc v= parameter and adds registration for that.

### CODE REVIEW

File: draft-ietf-quic-http.md [hunk 1]
```diff
@@ -97,38 +97,82 @@ defined in {{!RFC2119}}.
...
+the corresponding draft number to the identifier. For example, 
+draft-ietf-quic-http-01 is identified using the string "hq-01". 
+
+Non-compatible experiments that are based on these draft versions MUST append 
+the string "-" and an experiment name to the identifier. For example, an 
+experimental implementation based on draft-ietf-quic-http-09 which reserves an 
+extra stream for unsolicited transmission of 1980s pop music might identify
+itself as "hq-09-rickroll". Note that any label MUST conform to the "token"
+syntax defined in Section 3.2.6 of [RFC7230]. Experimenters are encouraged to
+coordinate their experiments on the ietf-quic-wg@w3.org mailing list.
```
Comment (ghedo) [inline draft-ietf-quic-http.md]: Should the mailing list be `quic@ietf.org`? The one listed above doesn't seem to exist https://lists.w3.org/Archives/Public/ietf-quic-wg/

Comment (MikeBishop) [inline draft-ietf-quic-http.md]: Bah, yes.  That's what I get for assuming mailing list norms stayed the same and copying text from the HTTP/2 spec.

File: draft-ietf-quic-http.md [hunk 2]
```diff
@@ -97,38 +97,82 @@ defined in {{!RFC2119}}.
...
+was available on UDP port 443 by including the following header in any
+response:
 
-In addition, the list of QUIC versions supported by the server can be specified
-by the v= parameter.  For example, if a server supported both version 33 and 34
-it would specify the following header:
+    Alt-Svc: hq=":443"
 
-   Alt-Svc: quic=":443"; v="34,33"
+## QUIC version hints {#alt-svc-version-hint}
```
Comment (martinthomson) [inline draft-ietf-quic-http.md]: Should we use "Title Case For Headers" or "Sentence case for headers"?

File: draft-ietf-quic-http.md [hunk 3]
```diff
@@ -97,38 +97,82 @@ defined in {{!RFC2119}}.
...
+QUIC versions are four-octet sequences with no additional constraints on format. 
+Octets not allowed in tokens ({{!RFC7230}}, Section 3.2.6) MUST be 
+percent-encoded as per Section 2.1 of {{!RFC3986}}. Consequently, the octet 
+representing the percent character ("%", hex 25) MUST be percent-encoded as 
+well. 
+
+On receipt of this header, a client MAY attempt to establish a QUIC connection
 on port 443 and, if successful, send HTTP requests using the mapping described
-in this document.
+in this document.  Servers SHOULD list only versions which they support, but MAY
```
Comment (martinthomson) [inline draft-ietf-quic-http.md]: This seems to refer to the example specifically.  but we're now a few paragraphs down.  Either speak generically "the client MAY attempt to establish a QUIC connection on the specified port" or refer to the example directly.

File: draft-ietf-quic-http.md [hunk 4]
```diff
@@ -95,40 +95,84 @@ It's not shouting; when they are capitalized, they have the special meaning
...
+version-negotiation hints to HTTP/QUIC clients. Syntax:
+
+    v = quoted-versions
+    quoted-versions = DQUOTE *( version-string "," ) version-string DQUOTE
+    version-string = token; percent-encoded QUIC version
+
+For example, if a server supported both version Q033 and Q034 it would specify
+the following header:
+
+    Alt-Svc: hq=":443"; v="Q034,Q033"
```
Comment (martinthomson) [inline draft-ietf-quic-http.md]: A note to discuss version representation:

You have a character string, I chose to go with just numbers.  Both have merits.  Numbers are simpler, but potentially longer (hexadecimal saves a few octets for larger values if you want to go there), but a four character string can be quite friendly.  You have to pick the right values though or you get a mash of %xx strings.

`Alt-Svc: hq=":443;v=<version>;v=<version>` might be better than creating a micro-syntax for the list.  See also httpwg/http-extensions#281

**Comment by mcmanus** [comment]
this is an important one to get in asap - thanks!

**Pull request merged**